osquery/tables/applications/linux/docker_deb_packages.cpp

```diff
@@ -0,0 +1,192 @@
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <sys/socket.h>
+#include <sys/types.h>
+
+#include <boost/algorithm/string.hpp>
+
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+#include <osquery/tables/applications/linux/namespace_ops.h>
+#include <osquery/tables/applications/posix/docker_api.h>
+#include <osquery/tables/system/linux/deb.h>
+
+#include <rapidjson/filereadstream.h>
+
+namespace osquery {
+namespace tables {
+
+static constexpr size_t BUF_SIZE{65536};
+
+void extractDebPackageInfo(const struct pkginfo* pkg,
+                           rapidjson::Writer<rapidjson::StringBuffer>& w) {
+  struct varbuf vb;
+  varbuf_init(&vb, 20);
+  w.StartObject();
+
+  // Iterate over the desired fieldinfos, calling their fwritefunctions
+  // to extract the package's information.
+  for (const struct fieldinfo& fip : fieldinfos) {
+    fip.wcall(&vb, pkg, &pkg->installed, fw_printheader, &fip);
+
+    std::string line = vb.string();
+    if (!line.empty()) {
+      size_t separator_position = line.find(':');
+
+      std::string key = line.substr(0, separator_position);
+      std::string value = line.substr(separator_position + 1, line.length());
+
+      auto it = kFieldMappings.find(key);
+      if (it != kFieldMappings.end()) {
+        boost::algorithm::trim(value);
+        w.Key(it->second.c_str());
+        w.String(std::move(value));
+      }
+    }
+    varbuf_reset(&vb);
+  }
+  varbuf_destroy(&vb);
+
+  w.EndObject();
+}
+
+void getDebPackages(int fd) {
+  if (!osquery::isDirectory(kDPKGPath)) {
+    TLOG << "Cannot find DPKG database: " << kDPKGPath;
+    return;
+  }
+
+  auto dropper = DropPrivileges::get();
+  dropper->dropTo("nobody");
+
+  rapidjson::StringBuffer buffer;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+  writer.StartArray();
+
+  struct pkg_array packages;
+  dpkg_setup(&packages);
+
+  const auto* pk = packages.pkgs;
+  for (const auto& pkg : boost::make_iterator_range(pk, pk + packages.n_pkgs)) {
+    // Casted to int to allow the older enums that were embeded in the
+    // packages
+    // struct to be compared
+    if (static_cast<int>(pkg->status) !=
+        static_cast<int>(PKG_STAT_NOTINSTALLED)) {
+      extractDebPackageInfo(pkg, writer);
+    }
+  }
+
+  dpkg_teardown(&packages);
+  writer.EndArray();
+
+  const auto bufSize = buffer.GetSize();
+  if (bufSize >= 1) {
+    if (write(fd, buffer.GetString(), bufSize) != (ssize_t)bufSize) {
+      VLOG(1) << "error writing pkginfo ";
+    }
+  }
+}
+
+QueryData genDockerDebPackages(QueryContext& context) {
+  QueryData results;
+
+  for (const auto& id : context.constraints["id"].getAll(EQUALS)) {
+    if (!checkConstraintValue(id)) {
+      continue;
+    }
+
+    pt::ptree container;
```

Comment:
I know you didn't introduce this in this PR so perhaps it's out of scope for this PR, but it would be nice if the docker API helper could be refactored one day to use a rapidjson data structure (or better: an api class with explicit methods and purpose built, type safe data structures based on the specific API requests).